src/shared/wall.c

```diff
@@ -0,0 +1,187 @@
+        const char *origin_tty,
+        bool (*match_tty)(const char *tty, bool is_local, void *userdata),
+        void *userdata) {
+
+#if ENABLE_UTMP
+        _unused_ _cleanup_(utxent_cleanup) bool utmpx = false;
+        struct utmpx *u;
+#endif
+        _cleanup_free_ char *text = NULL, *hn = NULL, *un = NULL, *stdin_tty = NULL;
+        int r;
+
+        hn = gethostname_malloc();
+        if (!hn)
+                return -ENOMEM;
+        if (!username) {
+                un = getlogname_malloc();
+                if (!un)
+                        return -ENOMEM;
+        }
+
+        if (!origin_tty) {
+                getttyname_harder(STDIN_FILENO, &stdin_tty);
+                origin_tty = stdin_tty;
+        }
+
+        if (asprintf(&text,
+                     "\r\n"
+                     "Broadcast message from %s@%s%s%s (%s):\r\n\r\n"
+                     "%s\r\n\r\n",
+                     un ?: username, hn,
+                     origin_tty ? " on " : "", strempty(origin_tty),
+                     FORMAT_TIMESTAMP(now(CLOCK_REALTIME)),
+                     message) < 0)
+                return -ENOMEM;
+
+        r = 0;
+
+#if ENABLE_UTMP
+        utmpx = utxent_start();
+
+        while ((u = getutxent())) {
+                _cleanup_free_ char *buf = NULL;
+                const char *path;
+                int q;
+
+                if (u->ut_type != USER_PROCESS || u->ut_user[0] == 0)
+                        continue;
+
+                /* This access is fine, because strlen("/dev/") < 32 (UT_LINESIZE) */
+                if (path_startswith(u->ut_line, "/dev/"))
+                        path = u->ut_line;
+                else {
+                        if (asprintf(&buf, "/dev/%.*s", (int) sizeof(u->ut_line), u->ut_line) < 0)
+                                return -ENOMEM;
+                        path = buf;
+                }
+
+                /* It seems that the address field is always set for remote logins.
+                 * For local logins and other local entries, we get [0,0,0,0]. */
+                bool is_local = memeqzero(u->ut_addr_v6, sizeof(u->ut_addr_v6));
+
+                if (!match_tty || match_tty(path, is_local, userdata)) {
+                        q = write_to_terminal(path, text);
+                        if (q < 0)
+                                r = q;
+                }
+        }
+#else
+        _cleanup_strv_free_ char **sessions = NULL;
+
+        r = sd_get_sessions(&sessions);
+        if (r < 0)
+                return r;
+
+        STRV_FOREACH(s, sessions) {
+                _cleanup_free_ char *buf = NULL, *tty = NULL, *rhost = NULL;
+                const char *path;
+                int q;
+
+                q = sd_session_get_tty(*s, &tty);
+                if (q < 0) {
+                        r = q;
+                        continue;
+                }
+
+                if (!tty)
+                        continue; /* ignore sessions without TTY */
```

Comment:
sd_session_get_tty() actually returns ENODATA in that case. which should hence be handled just like ENXIO: don't collect the failure code, just skip the session